src/TicketingSystem.PaymentsApi/Controllers/PaymentsController.cs

```diff
@@ -0,0 +1,124 @@
+        /// <returns></returns>
+        [HttpGet]
+        [Route("{paymentId}")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<IActionResult> GetPaymentStatus([FromRoute] string paymentId)
+        {
+            var payment = await _paymentService.GetByIdAsync(paymentId);
+
+            return payment is null ?
+                NotFound(nameof(paymentId))
+                : Ok(payment.State);
+        }
+
+        /// <summary>
+        /// Updates payment status and moves all the seats related to a payment to the sold state.
+        /// </summary>
+        /// <returns></returns>
+               [HttpPost]
+        [Route("{paymentId}/complete")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<IActionResult> CompletePayment([FromRoute] string paymentId)
+        {
+            var payment = await _paymentService.GetByIdAsync(paymentId);
+
+            if (payment is null)
+            {
+                return NotFound(nameof(paymentId));
+            }
+
+            var seatsUpdateStatus = await UpdatePaymentEventSeatsAsync(payment, EventSeatState.Sold);
+
+            await _paymentService.UpdatePaymentState(payment.Id, PaymentState.Completed);
+
+            return seatsUpdateStatus;
+        }
+
+        /// <summary>
+        /// Updates payment status and moves all the seats related to a payment to the sold state.
+        /// </summary>
+        /// <returns></returns>
+        [HttpPost]
+        [Route("{paymentId}/failed")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<IActionResult> FailPayment([FromRoute] string paymentId)
+        {
+            var payment = await _paymentService.GetByIdAsync(paymentId);
+
+            if (payment is null)
+            {
+                return NotFound(nameof(paymentId));
+            }
+
+            var seatsUpdateStatus = await UpdatePaymentEventSeatsAsync(payment, EventSeatState.Available);
+
+            await _paymentService.UpdatePaymentState(payment.Id, PaymentState.Failed);
+
+            return seatsUpdateStatus;
+        }
+
+        private async Task<IActionResult> UpdatePaymentEventSeatsAsync(PaymentDto payment, EventSeatState state)
+        {
+            var cartItems = await _cartItemService.FilterAsync(ci => ci.Id, payment.CartItemIds);
+
+            if (cartItems is null)
+            {
+                return BadRequest();
+            }
+
+            var eventSeats = await _eventSeatService.FilterAsync(es => es.Id, cartItems.Select(c => c.EventSeatId).ToList());
+
+            if (eventSeats is null || !eventSeats.Any())
+            {
```

Comment:
Again - additional redundant DB calls. If you put cartId and paymentId into eventSeat document with default state NULL and updating along with seatState with actions - you would get data faster.

Reply:
This is up to you - you can leave your current index models.